apps/crawler/lib/crawler/task_manager.ex

```diff
@@ -0,0 +1,148 @@
+defmodule Crawler.TaskManager do
+  @moduledoc """
+  Module to limit the amount of concurrent running task.
+  If the task maximum is reached, the request will be queued.
+  """
+
+  # struct to represent the inner gen_server state
+  defstruct [:max, :queue, :supervisor]
+
+  use GenServer
+  require Logger
+  alias Crawler
+
+  # macro to correct log debug messages
+  defmacrop debug(msg) do
```

Comment:
maybe an issue for a different pr, but this probably deserves its own module.

actually, this would be a good candidate for the new `common` app we plan on adding